### PR TITLE
Message Pool Rudimentary Spam Protection Measures

### PR DESCRIPTION
**Pull request by vyzo**
This adds tracking of required funds for pending messages so that we can stop spam attacks.
This also allows us to finally enable some score contribution from the messages topic.

### CODE REVIEW

**Comment by vyzo** [comment]
Follow up: we also want to rate limit messages per actor to some reasonable number.

**Comment by vyzo** [comment]
thanks for the review @schomatis!

> On the implementation side (but again with very little insight here), I'm not totally sold on the way this new requiredFunds state is maintained across calls, there seem to be many callers that now need to be aware of it and maintain it (which sounds a bit error prone to me).

Well, the state is maintained in the `add/rm` methods of the msgSet struct, and accesses with the `getRequiredFunds` method, so it should be safe.

**Comment by vyzo** [comment]
One potential follow up, probably for another pr because I want to think about it, is to turn the Ignores into Rejections if a peer is sending us too much stuff that gets Ignored.
Some kind of rate-limiting for resources we can consume from one peer -- but as I said, that's for another pr.

File: chain/messagepool/messagepool.go [hunk 1]
```diff
@@ -431,10 +491,6 @@ func (mp *MessagePool) addLocked(m *types.SignedMessage) error {
 		mp.blsSigCache.Add(m.Cid(), m.Signature)
 	}
 
-	if m.Message.GasLimit > build.BlockGasLimit {
-		return xerrors.Errorf("given message has too high of a gas limit")
-	}
-
```
Comment (vyzo) [inline chain/messagepool/messagepool.go:437]: this is checked by `ValidForBlockInclusion`, so it was redundant here and I removed it.

File: chain/messagepool/messagepool.go [hunk 2]
```diff
@@ -319,7 +348,7 @@ func (mp *MessagePool) checkMessage(m *types.SignedMessage) error {
 		return xerrors.Errorf("mpool message too large (%dB): %w", m.Size(), ErrMessageTooBig)
 	}
 
-	// Perform syntaxtic validation, minGas=0 as we check if correctly in select messages
+	// Perform syntactic validation, minGas=0 as we check the actual mingas before we add it
 	if err := m.Message.ValidForBlockInclusion(0); err != nil {
```
Comment (Kubuxu) [inline chain/messagepool/messagepool.go:352]: were do we do that check now?

Comment (vyzo) [reply chain/messagepool/messagepool.go:352]: we do it in both add and push.

Comment (vyzo) [reply chain/messagepool/messagepool.go:352]: and we check with the actual gas in the add logic, so it happens in all paths.

**Review by Kubuxu — APPROVED** [review]

File: chain/messagepool/messagepool.go [hunk 3]
```diff
@@ -150,12 +160,44 @@ func (ms *msgSet) add(m *types.SignedMessage, mp *MessagePool) (bool, error) {
...
+	ms.requiredFunds.Add(ms.requiredFunds, m.Message.RequiredFunds().Int)
+	//ms.requiredFunds.Add(ms.requiredFunds, m.Message.Value.Int)
 
 	return !has, nil
 }
 
+func (ms *msgSet) rm(nonce uint64) {
+	m, has := ms.msgs[nonce]
+	if has {
+		ms.requiredFunds.Sub(ms.requiredFunds, m.Message.RequiredFunds().Int)
```
Comment (whyrusleeping) [inline chain/messagepool/messagepool.go:183]: this should be `Add` no?

Comment (whyrusleeping) [reply chain/messagepool/messagepool.go:183]: Oh wait, we're tracking funds required for this message chain here, so this means we're removing the message from the chain, which lowers the required funds for that chain. SGTM

**Comment by vyzo** [comment]
I fixed a correctness issue by adding a flag in `msgSet.add/addLocked` to specify whether to check the limit or not.
This is to allow us to revert blocks safely even in the presence of a large number of pending messages (near or at the limit).

**Review by magik6k — APPROVED** [review]

**Pull request merged**